<a href="https://colab.research.google.com/github/unt-iialab/UNT-INFO5717-Fall2019/blob/master/Lesson%206/Lesson_six.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson Six: Classes and Objects

# Python Object Oriented Programming (OOP)


Python is a multi-paradigm programming language. Meaning, it supports different programming approach.

One of the popular approach to solve a programming problem is by creating objects. This is known as Object-Oriented Programming (OOP).

An object has two characteristics:

*   attributes
*   behavior

**Let's take an example:**

Parrot is an object,

*   name, age, color are attributes
*   singing, dancing are behavior







The concept of OOP in Python focuses on creating reusable code. This concept is also known as DRY (Don't Repeat Yourself).

In Python, the concept of OOP follows some basic principles:

| Principles   |      Description      |
|----------|:-------------:|
| Inheritance |  A process of using details from a new class without modifying existing class. | 
| Encapsulation |    Hiding the private details of a class from other objects. |
| Polymorphism | A concept of using common operation in different ways for different data input. |  



# What are classes and objects in Python?

As mentioned above, Python is an object oriented programming language. Unlike procedure oriented programming, where the main emphasis is on functions, object oriented programming stress on objects.

Object is simply a collection of data (variables) and methods (functions) that act on those data. And, class is a blueprint for the object.

We can think of **class as a sketch (prototype) of a house**. It contains all the details about the floors, doors, windows etc. Based on these descriptions we build the house. House is the object.

As, many houses can be made from a description, **we can create many objects from a class**. An object is also called an instance of a class and the process of creating this object is called instantiation.

**Example 1: Creating Class and Object in Python**


In [3]:
class Parrot:

    # class attribute
    species = "bird"

    # instance attribute
    def __init__(self, name, age):
        self.name = name
        self.age = age

# instantiate the Parrot class
blu = Parrot("Blu", 10)
woo = Parrot("Woo", 15)

# access the class attributes
print("Blu is a {}".format(blu.__class__.species))
print("Woo is also a {}".format(woo.__class__.species))

# access the instance attributes
print("{} is {} years old".format( blu.name, blu.age))
print("{} is {} years old".format( woo.name, woo.age))

Blu is a bird
Woo is also a bird
Blu is 10 years old
Woo is 15 years old


# Defining a Class in Python

Like function definitions begin with the keyword **def**, in Python, we define a class using the keyword **class**.

The first string is called docstring and has a brief description about the class. Although not mandatory, this is recommended.

Here is a simple class definition.


In [0]:
class MyNewClass:
    '''This is a docstring. I have created a new class'''
    pass

A class creates a new local namespace where all its attributes are defined. Attributes may be data or functions.

There are also special attributes in it that begins with double underscores (__). For example, __doc__ gives us the docstring of that class.

As soon as we define a class, a new class object is created with the same name. This class object allows us to access the different attributes as well as to instantiate new objects of that class.



**Example 2: Creating Class in Python**

In [5]:
class MyClass:
	"This is my second class"
	a = 10
	def func(self):
		print('Hello')

# Output: 10
print(MyClass.a)

# Output: <function MyClass.func at 0x0000000003079BF8>
print(MyClass.func)

# Output: 'This is my second class'
print(MyClass.__doc__)

10
<function MyClass.func at 0x7fbfef212f28>
This is my second class


# Creating an Object in Python

We saw that the class object could be used to access different attributes.

It can also be used to create new object instances (instantiation) of that class. The procedure to create an object is similar to a *function* call.


In [0]:
ob = MyClass()

This will create a new instance object named *ob*. We can access attributes of objects using the object name prefix.

Attributes may be data or method. Method of an object are corresponding functions of that class. Any function object that is a class attribute defines a method for objects of that class.

This means to say, since *MyClass.func* is a function object (attribute of class), *ob.func* will be a method object.



**Example 3: Creating Object in Python**

In [7]:
class MyClass:
	"This is my second class"
	a = 10
	def func(self):
		print('Hello')

# create a new MyClass
ob = MyClass()

# Output: <function MyClass.func at 0x000000000335B0D0>
print(MyClass.func)

# Output: <bound method MyClass.func of <__main__.MyClass object at 0x000000000332DEF0>>
print(ob.func)

# Calling function func()
# Output: Hello
ob.func()

<function MyClass.func at 0x7fbfee9b3488>
<bound method MyClass.func of <__main__.MyClass object at 0x7fbfee9bf710>>
Hello


You may have noticed the *self* parameter in function definition inside the class but, we called the method simply as *ob.func()* without any arguments. It still worked.

This is because, whenever an object calls its method, the object itself is passed as the first argument. So, *ob.func()* translates into *MyClass.func(ob)*.

# Methods in Python

Methods are functions defined inside the body of a class. They are used to define the behaviors of an object.

**Example 4: Creating Methods in Python**

In [8]:
class Parrot:
    
    # instance attributes
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    # instance method
    def sing(self, song):
        return "{} sings {}".format(self.name, song)

    def dance(self):
        return "{} is now dancing".format(self.name)

# instantiate the object
blu = Parrot("Blu", 10)

# call our instance methods
print(blu.sing("'Happy'"))
print(blu.dance())

Blu sings 'Happy'
Blu is now dancing


In general, calling a method with a list of n arguments is equivalent to calling the corresponding function with an argument list that is created by inserting the method's object before the first argument.

For these reasons, the first argument of the function in class must be the object itself. This is conventionally called self. It can be named otherwise but we highly recommend to follow the convention.

Now you must be familiar with class object, instance object, function object, method object and their differences.

# Constructors in Python

Class functions that begins with double underscore **(__)** are called special functions as they have special meaning.

Of one particular interest is the **__init__()** function. This special function gets called whenever a new object of that class is instantiated.

This type of function is also called constructors in Object Oriented Programming (OOP). We normally use it to initialize all the variables.

**Example 5: Creating constructors** 

In [9]:
class ComplexNumber:
    def __init__(self,r = 0,i = 0):
        self.real = r
        self.imag = i

    def getData(self):
        print("{0}+{1}j".format(self.real,self.imag))

# Create a new ComplexNumber object
c1 = ComplexNumber(2,3)

# Call getData() function
# Output: 2+3j
c1.getData()

# Create another ComplexNumber object
# and create a new attribute 'attr'
c2 = ComplexNumber(5)
c2.attr = 10

# Output: (5, 0, 10)
print((c2.real, c2.imag, c2.attr))

# but c1 object doesn't have attribute 'attr'
# AttributeError: 'ComplexNumber' object has no attribute 'attr'
c1.attr

2+3j
(5, 0, 10)


AttributeError: ignored

In the above example, we define a new class to represent complex numbers. It has two functions, **__init__()** to initialize the variables (defaults to zero) and **getData()** to display the number properly.

An interesting thing to note in the above step is that attributes of an object can be created on the fly. We created a new attribute **attr** for object **c2** and we read it as well. But this did not create that attribute for object **c1**.

# Deleting Attributes and Objects

Any attribute of an object can be deleted anytime, using the del statement. Try the following on the Python shell to see the output.



In [12]:
c1 = ComplexNumber(2,3)
del c1.imag
c1.getData()

AttributeError: ignored

In [13]:
del ComplexNumber.getData
c1.getData()

AttributeError: ignored

We can even delete the object itself, using the del statement.

In [14]:
c1 = ComplexNumber(1,3)
del c1
c1

NameError: ignored

Actually, it is more complicated than that. When we do *c1 = ComplexNumber(1,3)*, a new instance object is created in memory and the name *c1* binds with it.

On the command del c1, this binding is removed and the name *c1* is deleted from the corresponding namespace. The object however continues to exist in memory and if no other name is bound to it, it is later automatically destroyed.

This automatic destruction of unreferenced objects in Python is also called garbage collection.

![alt text](https://cdn.programiz.com/sites/tutorial2program/files/objectReference.jpg)

# Inheritance in Python 

Inheritance is a way of creating new class for using details of existing class without modifying it. The newly formed class is a derived class (or child class). Similarly, the existing class is a base class (or parent class).

**Example 6: Use of Inheritance in Python**


In [15]:
# parent class
class Bird:
    
    def __init__(self):
        print("Bird is ready")

    def whoisThis(self):
        print("Bird")

    def swim(self):
        print("Swim faster")

# child class
class Penguin(Bird):

    def __init__(self):
        # call super() function
        super().__init__()
        print("Penguin is ready")

    def whoisThis(self):
        print("Penguin")

    def run(self):
        print("Run faster")

peggy = Penguin()
peggy.whoisThis()
peggy.swim()
peggy.run()

Bird is ready
Penguin is ready
Penguin
Swim faster
Run faster


In the above program, we created two classes i.e. *Bird* (parent class) and *Penguin* (child class). The child class inherits the functions of parent class. We can see this from *swim()* method. Again, the child class modified the behavior of parent class. We can see this from whoisThis() method. Furthermore, we extend the functions of parent class, by creating a new *run()* method.

Additionally, we use *super()* function before *__init__()* method. This is because we want to pull the content of *__init__()* method from the parent class into the child class.

# Encapsulation in Python 

Using OOP in Python, we can restrict access to methods and variables. This prevent data from direct modification which is called encapsulation. In Python, we denote private attribute using underscore as prefix i.e single “ _ “ or double “ __“.

**Example 7: Data Encapsulation in Python**

In [16]:
class Computer:

    def __init__(self):
        self.__maxprice = 900

    def sell(self):
        print("Selling Price: {}".format(self.__maxprice))

    def setMaxPrice(self, price):
        self.__maxprice = price

c = Computer()
c.sell()

# change the price
c.__maxprice = 1000
c.sell()

# using setter function
c.setMaxPrice(1000)
c.sell()

Selling Price: 900
Selling Price: 900
Selling Price: 1000


In the above program, we defined a class Computer. We use __init__() method to store the maximum selling price of computer. We tried to modify the price. However, we can’t change it because Python treats the __maxprice as private attributes. To change the value, we used a setter function i.e setMaxPrice() which takes price as parameter.

# Polymorphism in Python 

Polymorphism is an ability (in OOP) to use common interface for multiple form (data types).

Suppose, we need to color a shape, there are multiple shape option (rectangle, square, circle). However we could use same method to color any shape. This concept is called Polymorphism.

In [17]:
class Parrot:

    def fly(self):
        print("Parrot can fly")
    
    def swim(self):
        print("Parrot can't swim")

class Penguin:

    def fly(self):
        print("Penguin can't fly")
    
    def swim(self):
        print("Penguin can swim")

# common interface
def flying_test(bird):
    bird.fly()

#instantiate objects
blu = Parrot()
peggy = Penguin()

# passing the object
flying_test(blu)
flying_test(peggy)

Parrot can fly
Penguin can't fly


In the above program, we defined two classes *Parrot* and *Penguin*. Each of them have common method *fly()* method. However, their functions are different. To allow polymorphism, we created common interface i.e *flying_test()* function that can take any object. Then, we passed the objects *blu* and *peggy* in the *flying_test()* function, it ran effectively.

# Advantages of Object Oriented Programming (OOP)


*   The programming gets easy and efficient.
*   The class is sharable, so codes can be reused.
*   The productivity of programmars increases.
*   Data is safe and secure with data abstraction.





# Examples from the textbook

**`Code examples from this chapter 15`**

In [22]:
"""

Code example from Think Python, by Allen B. Downey.
Available from http://thinkpython.com

Copyright 2012 Allen B. Downey.
Distributed under the GNU General Public License at gnu.org/licenses/gpl.html.

"""

class Point(object):
    """Represents a point in 2-D space."""


def print_point(p):
    """Print a Point object in human-readable format."""
    print ('(%g, %g)' % (p.x, p.y))


class Rectangle(object):
    """Represents a rectangle. 

    attributes: width, height, corner.
    """


def find_center(rect):
    """Returns a Point at the center of a Rectangle."""
    p = Point()
    p.x = rect.corner.x + rect.width/2.0
    p.y = rect.corner.y + rect.height/2.0
    return p


def grow_rectangle(rect, dwidth, dheight):
    """Modify the Rectangle by adding to its width and height.

    rect: Rectangle object.
    dwidth: change in width (can be negative).
    dheight: change in height (can be negative).
    """
    rect.width += dwidth
    rect.height += dheight


def main():
    blank = Point()
    blank.x = 3
    blank.y = 4
    print ('blank',)
    print_point(blank)

    box = Rectangle()
    box.width = 100.0
    box.height = 200.0
    box.corner = Point()
    box.corner.x = 0.0
    box.corner.y = 0.0

    center = find_center(box)
    print ('center',)
    print_point(center)

    print (box.width)
    print (box.height)
    print ('grow')
    grow_rectangle(box, 50, 100)
    print (box.width)
    print (box.height)


if __name__ == '__main__':
    main()


blank
(3, 4)
center
(50, 100)
100.0
200.0
grow
150.0
300.0


In [23]:
"""This module contains code from
Think Python by Allen B. Downey
http://thinkpython.com

Copyright 2012 Allen B. Downey
License: GNU GPLv3 http://www.gnu.org/licenses/gpl.html

"""

import copy
import math

# to avoid duplicating code, I'm importing everything from Point1 
from Point1 import *


def distance_between_points(p1, p2):
    """Computes the distance between two Point objects."""
    dx = p1.x - p2.x
    dy = p1.y - p2.y
    dist = math.sqrt(dx**2 + dy**2)
    return dist


def move_rectangle(rect, dx, dy):
    """Move the Rectangle by modifying its corner object.

    rect: Rectangle object.
    dx: change in x coordinate (can be negative).
    dy: change in y coordinate (can be negative).
    """
    rect.corner.x += dx
    rect.corner.y += dy


def move_rectangle_copy(rect, dx, dy):
    """Move the Rectangle and return a new Rectangle object.

    rect: Rectangle object.
    dx: change in x coordinate (can be negative).
    dy: change in y coordinate (can be negative).
    """
    new = copy.deepcopy(rect)
    move_rectangle(new, dx, dy)
    return new


def main():
    blank = Point()
    blank.x = 0
    blank.y = 0

    grosse = Point()
    grosse.x = 3
    grosse.y = 4

    print ('distance'+',')
    print (distance_between_points(grosse, blank))


    box = Rectangle()
    box.width = 100.0
    box.height = 200.0
    box.corner = Point()
    box.corner.x = 50.0
    box.corner.y = 50.0

    print (box.corner.x)
    print (box.corner.y)
    print ('move')
    move_rectangle(box, 50, 100)
    print (box.corner.x)
    print (box.corner.y)

    new_box = move_rectangle_copy(box, 50, 100)
    print (box.corner.x)
    print (box.corner.y)

if __name__ == '__main__':
    main()


distance,
5.0
50.0
50.0
move
100.0
150.0
100.0
150.0


**Exercise 15.1.** 

Write a definition for a class named Circle with attributes center and radius,
where center is a Point object and radius is a number.

Instantiate a Circle object that represents a circle with its center at (150, 100) and radius 75.

Write a function named point_in_circle that takes a Circle and a Point and returns True if the
Point lies in or on the boundary of the circle.

Write a function named rect_in_circle that takes a Circle and a Rectangle and returns True if
the Rectangle lies entirely in or on the boundary of the circle.

Write a function named rect_circle_overlap that takes a Circle and a Rectangle and returns
True if any of the corners of the Rectangle fall inside the circle. Or as a more challenging version,
return True if any part of the Rectangle falls inside the circle.

In [25]:
"""This module contains a code example related to

Think Python, 2nd Edition
by Allen Downey
http://thinkpython2.com

Copyright 2015 Allen Downey

License: http://creativecommons.org/licenses/by/4.0/
"""

from __future__ import print_function, division

import copy

from Point1 import Point, Rectangle, print_point
from Point1_soln import distance_between_points


class Circle:
    """Represents a circle.

    Attributes: center, radius
    """


def point_in_circle(point, circle):
    """Checks whether a point lies inside a circle (or on the boundary).

    point: Point object
    circle: Circle object
    """
    d = distance_between_points(point, circle.center)
    print(d)
    return d <= circle.radius


def rect_in_circle(rect, circle):
    """Checks whether the corners of a rect fall in/on a circle.

    rect: Rectangle object
    circle: Circle object
    """
    p = copy.copy(rect.corner)
    print_point(p)
    if not point_in_circle(p, circle):
        return False

    p.x += rect.width
    print_point(p)
    if not point_in_circle(p, circle):
        return False

    p.y -= rect.height
    print_point(p)
    if not point_in_circle(p, circle):
        return False

    p.x -= rect.width
    print_point(p)
    if not point_in_circle(p, circle):
        return False

    return True


def rect_circle_overlap(rect, circle):
    """Checks whether any corners of a rect fall in/on a circle.

    rect: Rectangle object
    circle: Circle object
    """
    p = copy.copy(rect.corner)
    print_point(p)
    if point_in_circle(p, circle):
        return True

    p.x += rect.width
    print_point(p)
    if point_in_circle(p, circle):
        return True

    p.y -= rect.height
    print_point(p)
    if point_in_circle(p, circle):
        return True

    p.x -= rect.width
    print_point(p)
    if point_in_circle(p, circle):
        return True

    return False


def main():
    box = Rectangle()
    box.width = 100.0
    box.height = 200.0
    box.corner = Point()
    box.corner.x = 50.0
    box.corner.y = 50.0

    print(box.corner.x)
    print(box.corner.y)

    circle = Circle
    circle.center = Point()
    circle.center.x = 150.0
    circle.center.y = 100.0
    circle.radius = 75.0

    print(circle.center.x)
    print(circle.center.y)
    print(circle.radius)

    print(point_in_circle(box.corner, circle))
    print(rect_in_circle(box, circle))
    print(rect_circle_overlap(box, circle))


if __name__ == '__main__':
    main()

50.0
50.0
150.0
100.0
75.0
111.80339887498948
False
(50, 50)
111.80339887498948
False
(50, 50)
111.80339887498948
(150, 50)
50.0
True


**Code examples from this chapter 16**

In [37]:
"""

Code example from Think Python, by Allen B. Downey.
Available from http://thinkpython.com

Copyright 2012 Allen B. Downey.
Distributed under the GNU General Public License at gnu.org/licenses/gpl.html.

"""

class Time(object):
    """Represents the time of day.
       
    attributes: hour, minute, second
    """

def print_time(t):
    print ('%.2d:%.2d:%.2d' % (t.hour, t.minute, t.second))


def int_to_time(seconds):
    """Makes a new Time object.

    seconds: int seconds since midnight.
    """
    time = Time()
    minutes, time.second = divmod(seconds, 60)
    time.hour, time.minute = divmod(minutes, 60)
    return time


def time_to_int(time):
    """Computes the number of seconds since midnight.

    time: Time object.
    """
    minutes = time.hour * 60 + time.minute
    seconds = minutes * 60 + time.second
    return seconds


def add_times(t1, t2):
    """Adds two time objects."""
    assert valid_time(t1) and valid_time(t2)
    seconds = time_to_int(t1) + time_to_int(t2)
    return int_to_time(seconds)


def valid_time(time):
    """Checks whether a Time object satisfies the invariants."""
    if time.hour < 0 or time.minute < 0 or time.second < 0:
        return False
    if time.minute >= 60 or time.second >= 60:
        return False
    return True


def main():
    # if a movie starts at noon...
    noon_time = Time()
    noon_time.hour = 12
    noon_time.minute = 0
    noon_time.second = 0

    print ('Starts at',)
    print_time(noon_time)

    # and the run time of the movie is 109 minutes...
    movie_minutes = 109
    run_time = int_to_time(movie_minutes * 60)
    print ('Run time',)
    print_time(run_time)

    # what time does the movie end?
    end_time = add_times(noon_time, run_time)
    print ('Ends at',)
    print_time(end_time)

if __name__ == '__main__':
    main()

Starts at
12:00:00
Run time
01:49:00
Ends at
13:49:00


**Exercise 16.1.** 

Write a function called **mul_time** that takes *a Time object* and *a number* and *returns*
*a new Time object that contains the product of the original Time and the number*.
Then use mul_time to write a function that takes a Time object that represents the finishing time
in a race, and a number that represents the distance, and returns a Time object that represents the
average pace (time per mile).

In [41]:
def mul_time(t1, factor):
    """Multiplies a Time object by a factor."""
    assert valid_time(t1)
    seconds = time_to_int(t1) * factor
    return int_to_time(seconds)

race_time = Time()
race_time.hour = 1
race_time.minute = 34
race_time.second = 5

print('Half marathon time', end=' ')
print_time(race_time)

distance = 13.1       # miles
pace = mul_time(race_time, 1/distance)

print('Time per mile', end=' ')
print_time(pace)

Half marathon time 01:34:05
Time per mile 00:07:10


**Exercise 16.2.** 

The datetime module provides time objects that are similar to the Time objects
in this chapter, but they provide a rich set of methods and operators. 

1. Use the datetime module to write a program that gets the current date and prints the day of
the week.
2. Write a program that takes a birthday as input and prints the user’s age and the number of
days, hours, minutes and seconds until their next birthday.
3. For two people born on different days, there is a day when one is twice as old as the other.
That’s their Double Day. Write a program that takes two birth dates and computes their
Double Day.
4. For a little more challenge, write the more general version that computes the day when one
person is n times older than the other

In [38]:
"""This module contains a code example related to

Think Python, 2nd Edition
by Allen Downey
http://thinkpython2.com

Copyright 2015 Allen Downey

License: http://creativecommons.org/licenses/by/4.0/
"""

from __future__ import print_function, division

from datetime import datetime


def main():
    print("Today's date and the day of the week:")
    today = datetime.today()
    print(today)
    print(today.strftime("%A"))

    print("Your next birthday and how far away it is:")
    #s = input('Enter your birthday in mm/dd/yyyy format: ')
    s = '5/11/1967'
    bday = datetime.strptime(s, '%m/%d/%Y')

    next_bday = bday.replace(year=today.year)
    if next_bday < today:
        next_bday = next_bday.replace(year=today.year+1)
    print(next_bday)

    until_next_bday = next_bday - today
    print(until_next_bday)

    print("Your current age:")
    last_bday = next_bday.replace(year=next_bday.year-1)
    age = last_bday.year - bday.year
    print(age)

    print("For people born on these dates:")
    bday1 = datetime(day=11, month=5, year=1967)
    bday2 = datetime(day=11, month=10, year=2003)
    print(bday1)
    print(bday2)

    print("Double Day is")
    d1 = min(bday1, bday2)
    d2 = max(bday1, bday2)
    dd = d2 + (d2 - d1)
    print(dd)


if __name__ == '__main__':
    main()

Today's date and the day of the week:
2019-09-28 20:32:39.583470
Saturday
Your next birthday and how far away it is:
2020-05-11 00:00:00
225 days, 3:27:20.416530
Your current age:
52
For people born on these dates:
1967-05-11 00:00:00
2003-10-11 00:00:00
Double Day is
2040-03-12 00:00:00


**Code examples from this chapter 16**

In [42]:
"""This module contains a code example related to

Think Python, 2nd Edition
by Allen Downey
http://thinkpython2.com

Copyright 2015 Allen Downey

License: http://creativecommons.org/licenses/by/4.0/
"""

from __future__ import print_function, division


class Time:
    """Represents the time of day.
       
    attributes: hour, minute, second
    """
    def __init__(self, hour=0, minute=0, second=0):
        """Initializes a time object.

        hour: int
        minute: int
        second: int or float
        """
        self.hour = hour
        self.minute = minute
        self.second = second

    def __str__(self):
        """Returns a string representation of the time."""
        return '%.2d:%.2d:%.2d' % (self.hour, self.minute, self.second)

    def print_time(self):
        """Prints a string representation of the time."""
        print(str(self))

    def time_to_int(self):
        """Computes the number of seconds since midnight."""
        minutes = self.hour * 60 + self.minute
        seconds = minutes * 60 + self.second
        return seconds

    def is_after(self, other):
        """Returns True if t1 is after t2; false otherwise."""
        return self.time_to_int() > other.time_to_int()

    def __add__(self, other):
        """Adds two Time objects or a Time object and a number.

        other: Time object or number of seconds
        """
        if isinstance(other, Time):
            return self.add_time(other)
        else:
            return self.increment(other)

    def __radd__(self, other):
        """Adds two Time objects or a Time object and a number."""
        return self.__add__(other)

    def add_time(self, other):
        """Adds two time objects."""
        assert self.is_valid() and other.is_valid()
        seconds = self.time_to_int() + other.time_to_int()
        return int_to_time(seconds)

    def increment(self, seconds):
        """Returns a new Time that is the sum of this time and seconds."""
        seconds += self.time_to_int()
        return int_to_time(seconds)

    def is_valid(self):
        """Checks whether a Time object satisfies the invariants."""
        if self.hour < 0 or self.minute < 0 or self.second < 0:
            return False
        if self.minute >= 60 or self.second >= 60:
            return False
        return True


def int_to_time(seconds):
    """Makes a new Time object.

    seconds: int seconds since midnight.
    """
    minutes, second = divmod(seconds, 60)
    hour, minute = divmod(minutes, 60)
    time = Time(hour, minute, second)
    return time


def main():
    start = Time(9, 45, 00)
    start.print_time()

    end = start.increment(1337)
    #end = start.increment(1337, 460)
    end.print_time()

    print('Is end after start?')
    print(end.is_after(start))

    print('Using __str__')
    print(start, end)

    start = Time(9, 45)
    duration = Time(1, 35)
    print(start + duration)
    print(start + 1337)
    print(1337 + start)

    print('Example of polymorphism')
    t1 = Time(7, 43)
    t2 = Time(7, 41)
    t3 = Time(7, 37)
    total = sum([t1, t2, t3])
    print(total)


if __name__ == '__main__':
    main()

09:45:00
10:07:17
Is end after start?
True
Using __str__
09:45:00 10:07:17
11:20:00
10:07:17
10:07:17
Example of polymorphism
23:01:00


**Exercise 18.2.** 

Write a Deck method called deal_hands that takes two parameters, the number of
hands and the number of cards per hand. It should create the appropriate number of Hand objects,
deal the appropriate number of cards per hand, and return a list of Hands.

In [51]:
import random

class Card:
	''' class represents suit of cars, encoding with numbers 
	
	Attributes:
		suit;
		rank;
	
	sapdes -> 3
	hearts -> 2
	diamonds -> 1
	clubs -> 0
	
	jack -> 11
	queen -> 12
	king -> 13
	
	
	'''
	# Attribute
	suit_names = ['Clubs', 'Diamonds', 'Hearts', 'Spades']
	rank_names = [None, 'Ace', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack', 'Queen', 'King']
	
	def __init__(self, suit=0, rank=2):
		self.suit = suit
		self.rank = rank
		
	def __str__(self):
		return '%s of %s' % (Card.rank_names[self.rank], Card.suit_names[self.suit])
	
	def __lt__(self, other):
		# first compare suit
		if self.suit < other.suit: return True
		if self.suit > other.suit: return False
		
		# if suit is the same then compare rank
		return self.rank < other.rank
		
		
class Deck:
	'''class represents deck
	
	Attributes:
		cards: list
	'''
	
	def __init__(self):
		self.cards = []
		for suit in range(4):
			for rank in range(1, 14):
				card = Card(suit, rank)
				self.cards.append(card)
	
	def __str__(self):
		res = []
		for card in self.cards:
			res.append(str(card))
		return '\n'.join(res)
		
	def pop_card(self):
		'''remove the last card from the deck'''
		return self.cards.pop()
		
	def add_card(self, card):
		''' add a card to the last position of deck (using veneer methods)'''
		self.cards.append(card)
		
	def shuffle(self):
		'''shuffle the deck randomly'''
		random.shuffle(self.cards)
		
	def sort(self):
		'''sort the deck in descending ordoer'''
		self.cards.sort()
		
	def move_card(self, hand, num):
		'''move card(s) to another Hand or Deck'''
		for i in range(num):
			hand.add_card(self.pop_card())
			
	def deal_hands(self, num_of_hands, num_of_cards_per_hand):
		'''deal hands
		
		num_of_hands: int
		num_of_cards_per_hand: list
		
		'''
		hand_list = []
		for i in range(num_of_hands):
			h = Hand("Hand No. %d" % i)
			self.move_card(h, num_of_cards_per_hand)
			hand_list.append(h)
		return hand_list
		
		
class Hand(Deck):
	'''class represents cards on hand, inherited from deck '''
	
	def __init__(self, label=''):
		'''initialize cards on hand, and label which hand is using'''
		self.cards = []
		self.label = label
		
if __name__ == '__main__':
	king_of_spades = Card(3, 13) # test
	print(king_of_spades)
	d = Deck() #test
	print('\n---initial result---')
	print(d)
	d.shuffle()
	print('\n---shuffle result---')
	print(d)
	d.sort()
	print('\n---sort result---')
	print(d)
	print('\n---deal_hand test result---')
	li = d.deal_hands(5,4)
	for h in li:
		print (h.label)
		print (h)

King of Spades

---initial result---
Ace of Clubs
2 of Clubs
3 of Clubs
4 of Clubs
5 of Clubs
6 of Clubs
7 of Clubs
8 of Clubs
9 of Clubs
10 of Clubs
Jack of Clubs
Queen of Clubs
King of Clubs
Ace of Diamonds
2 of Diamonds
3 of Diamonds
4 of Diamonds
5 of Diamonds
6 of Diamonds
7 of Diamonds
8 of Diamonds
9 of Diamonds
10 of Diamonds
Jack of Diamonds
Queen of Diamonds
King of Diamonds
Ace of Hearts
2 of Hearts
3 of Hearts
4 of Hearts
5 of Hearts
6 of Hearts
7 of Hearts
8 of Hearts
9 of Hearts
10 of Hearts
Jack of Hearts
Queen of Hearts
King of Hearts
Ace of Spades
2 of Spades
3 of Spades
4 of Spades
5 of Spades
6 of Spades
7 of Spades
8 of Spades
9 of Spades
10 of Spades
Jack of Spades
Queen of Spades
King of Spades

---shuffle result---
8 of Clubs
2 of Spades
2 of Diamonds
4 of Hearts
Ace of Diamonds
7 of Diamonds
6 of Diamonds
10 of Clubs
10 of Hearts
Jack of Clubs
Jack of Hearts
7 of Clubs
10 of Spades
5 of Hearts
King of Spades
Queen of Spades
3 of Hearts
Jack of Diamonds
2 of Club

**Exercise 18.3.** 

The following are the possible hands in poker, in increasing order of value and
decreasing order of probability:

**pair:** two cards with the same rank

**two pair:** two pairs of cards with the same rank

**three of a kind:** three cards with the same rank

**straight:** five cards with ranks in sequence (aces can be high or low, so Ace-2-3-4-5 is a straight
and so is 10-Jack-Queen-King-Ace, but Queen-King-Ace-2-3 is not.)

**flush:** five cards with the same suit

**full house:** three cards with one rank, two cards with another

**four of a kind:** four cards with the same rank

**straight flush:** five cards in sequence (as defined above) and with the same suit

In [53]:
"""Adapeted from Think Python website
Copyright 2017 Dex D. Hunter
License: http://creativecommons.org/licenses/by/4.0/
"""

from __future__ import print_function, division

from Card import Hand, Deck


class PokerHand(Hand):
	"""Represents a poker hand.
	
	Inherited from Hand
	"""

	def suit_hist(self):
		"""Builds a histogram of the suits that appear in the hand.
		Stores the result in attribute suits.
		"""
		self.suits = {}
		for card in self.cards:
			self.suits[card.suit] = self.suits.get(card.suit, 0) + 1
		print(self.suits)
		
	def rank_hist(self):
		"""Builds a histogram of the rank that appear in the hand.
		Stores the result in attribute rank.
		"""
		self.ranks = {}
		for card in self.cards:
			self.ranks[card.rank] = self.ranks.get(card.rank, 0) + 1
		print(self.ranks)

	def has_flush(self):
		"""Returns True if the hand has a flush, False otherwise.
	  
		Note that this works correctly for hands with more than 5 cards.
		
		flush: 5 cards with the same suit
		"""
		self.suit_hist()
		for val in self.suits.values():
			if val >= 5:
				return True
		return False
		
	def has_pair(self, other):
		'''pair: two cards with the same rank'''
		self.rank_hist()
		for val in self.ranks.values():
			if val >= 2:
				return True
		return False
		
	
	def has_twopair(self):
		''' has two pairs of cards with the same rank '''
		self.rank_hist()
		count = 0
		for val in self.ranks.values():
			if val >= 2:
				count += 1
				
		if count >=2:
			return True
		return False
		
	def classify(self):
		'''
		straight: five cards with ranks in sequence
		
		flush: five cards with the same suit
		'''
		label = ['pair', 'two pair', 'three of a kind', 'striaght', 'flush', 'full house', 'four of a kind', 'straight flush']
		


if __name__ == '__main__':
	# make a deck
	deck = Deck()
	deck.shuffle()

	# deal the cards and classify the hands
	h = PokerHand()
	deck.move_card(h, 10)
	h.suit_hist()
	print(h)
	'''
	for i in range(7):
		hand = PokerHand()
		deck.move_cards(hand, 7)
		hand.suit_hist
		hand.sort()
		print(hand)
		print(hand.has_flush())
		print('')
	'''

{1: 2, 2: 3, 0: 2, 3: 3}
Queen of Diamonds
9 of Hearts
Ace of Hearts
2 of Clubs
9 of Spades
5 of Clubs
Ace of Spades
8 of Diamonds
Queen of Spades
2 of Hearts
